system/lib/wasmfs/backends/opfs_backend.cpp

```
@@ -101,92 +114,221 @@ namespace {
-    assert(accessID == -1);
+    assert(state.getKind() == OpenState::None);
     proxy([&]() { _wasmfs_opfs_free_file(fileID); });
   }
 
 private:
   size_t getSize() override {
+    // TODO: 64-bit sizes.
     uint32_t size;
-    if (accessID == -1) {
-      proxy(
-        [&](auto ctx) { _wasmfs_opfs_get_size_file(ctx.ctx, fileID, &size); });
-    } else {
-      proxy([&](auto ctx) {
-        _wasmfs_opfs_get_size_access(ctx.ctx, accessID, &size);
-      });
+    switch (state.getKind()) {
+      case OpenState::None:
+        proxy([&](auto ctx) {
+          _wasmfs_opfs_get_size_file(ctx.ctx, fileID, &size);
+        });
+        break;
+      case OpenState::Access:
+        proxy([&](auto ctx) {
+          _wasmfs_opfs_get_size_access(ctx.ctx, state.getAccessID(), &size);
+        });
+        break;
+      case OpenState::Blob:
+        proxy([&]() { size = _wasmfs_opfs_get_size_blob(state.getBlobID()); });
+        break;
+      default:
+        WASMFS_UNREACHABLE("Unexpected open state");
     }
     return size_t(size);
   }
 
   void setSize(size_t size) override {
-    if (accessID == -1) {
-      int err;
-      proxy([&](auto ctx) {
-        _wasmfs_opfs_set_size_file(ctx.ctx, fileID, size, &err);
-      });
-      assert(err == 0 && "TODO: better error handling");
-    } else {
-      proxy([&](auto ctx) {
-        _wasmfs_opfs_set_size_access(ctx.ctx, accessID, size);
-      });
+    switch (state.getKind()) {
+      case OpenState::Access:
+        proxy([&](auto ctx) {
+          _wasmfs_opfs_set_size_access(ctx.ctx, state.getAccessID(), size);
+        });
+        break;
+      case OpenState::Blob:
+        // We don't support `truncate` in blob mode because the blob would
+        // become invalidated and refreshing it while ensuring other in-flight
+        // operations on the same file do not observe the invalidated blob would
+        // be extermely complicated.
+        assert(false && "TODO: proper error handling");
```

I worry about depending on an assertion here, as if at some point we ship a build without assertions it would silently fall through, and the results could be very weird. How about `WASMFS_UNREACHABLE`?